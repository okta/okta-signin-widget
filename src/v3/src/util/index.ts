/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export * from './browserUtils';
export * from './buildErrorMessageIds';
export * from './clipboard';
export * from './cookieUtils';
export * from './environmentUtils';
export * from './flattenInputs';
export * from './formatError';
export * from './formUtils';
export * from './generateRandomString';
export * from './getAuthenticatorKey';
export * from './getCurrentAuthenticator';
export * from './getElementKey';
export * from './getImmutableData';
export * from './getTranslation';
export * from './getValidationMessages';
export * from './idxUtils';
export * from './isInteractiveElement';
export * from './isPasswordRecovery';
export * from './isPollingStep';
export * from './languageUtils';
export * from './locUtil';
export * from './passwordUtils';
export * from './removeFieldLevelMessages';
export * from './resetMessagesToInputs';
export * from './settingsUtils';
export * from './shouldShowCancelLink';
export * from './toNestedObject';
export * from './webauthnUtils';
