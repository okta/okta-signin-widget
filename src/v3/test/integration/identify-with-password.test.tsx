/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { waitFor } from '@testing-library/preact';
import { createAuthJsPayloadArgs, setup } from './util';

import * as cookieUtils from '../../src/util/cookieUtils';
import mockResponse from '../../src/mocks/response/idp/idx/introspect/default.json';
import { WidgetOptions } from '../../../types';

describe('identify-with-password', () => {
  it('should display Identifier & Password hint labels', async () => {
    const usernameHint = 'This is your username';
    const passwordHint = 'This is your password';
    const hintOverrides: WidgetOptions['i18n'] = {
      en: {
        'primaryauth.username.tooltip': usernameHint,
        'primaryauth.password.tooltip': passwordHint,
      },
    };
    const {
      findByTestId, findByText,
    } = await setup({
      mockResponse,
      widgetOptions: {
        i18n: hintOverrides,
      },
    });

    await findByText('Sign in', { selector: 'button' });
    const usernameEl = await findByTestId('identifier-hint') as HTMLLabelElement;
    const passwordEl = await findByTestId('credentials.passcode-hint') as HTMLLabelElement;

    expect(usernameEl.textContent).toEqual(usernameHint);
    expect(passwordEl.textContent).toEqual(passwordHint);
  });

  it('renders the loading state first', async () => {
    const { container, findByLabelText } = await setup({ mockResponse });
    await findByLabelText('Processing...');
    expect(container).toMatchSnapshot();
  });

  it('renders form with focus', async () => {
    const { container, findByLabelText } = await setup({ mockResponse });
    const inputEle = await findByLabelText(/Username/) as HTMLInputElement;
    expect(container).toMatchSnapshot();
    expect(inputEle).not.toHaveFocus();
  });

  it('renders form without focus', async () => {
    const { container, findByLabelText } = await setup({
      mockResponse, widgetOptions: { features: { autoFocus: true } },
    });
    const inputEle = await findByLabelText(/Username/) as HTMLInputElement;
    expect(container).toMatchSnapshot();
    await waitFor(() => expect(inputEle).toHaveFocus());
  });

  it('should pre-populate username into identifier field when set in widget config props', async () => {
    const mockUsername = 'testuser@okta1.com';
    const {
      findByTestId,
    } = await setup({
      mockResponse,
      widgetOptions: { username: mockUsername },
    });

    const usernameEl = await findByTestId('identifier') as HTMLInputElement;

    expect(usernameEl.value).toBe(mockUsername);
  });

  it('should pre-populate username into identifier field when set in cookie', async () => {
    const mockUsername = 'testuser@okta1.com';
    jest.spyOn(cookieUtils, 'getUsernameCookie').mockReturnValue(mockUsername);
    const {
      findByTestId,
    } = await setup({
      mockResponse,
      widgetOptions: { features: { rememberMe: true } },
    });

    const usernameEl = await findByTestId('identifier') as HTMLInputElement;

    expect(usernameEl.value).toBe(mockUsername);
  });

  describe('Client-side field change validation tests', () => {
    it('should attempt to submit page with all required fields empty', async () => {
      const {
        authClient,
        user,
        container,
        findByTestId,
        findByText,
      } = await setup({ mockResponse, widgetOptions: { features: { autoFocus: true } } });

      await findByTestId('identifier') as HTMLInputElement;
      await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByText('Sign in', { selector: 'button' });

      await user.click(submitButton);
      await findByText(/We found some errors./);
      const identifierError = await findByTestId('identifier-error');
      const passwordError = await findByTestId('credentials.passcode-error');

      expect(identifierError.textContent).toEqual('This field cannot be left blank');
      expect(container).toMatchSnapshot();
      expect(passwordError.textContent).toEqual('This field cannot be left blank');
      expect(authClient.options.httpRequestClient).not.toHaveBeenCalled();
    });

    it('should attempt to submit page with all required fields empty and type in one of the required fields to remove error', async () => {
      const {
        authClient,
        user,
        container,
        findByTestId,
        queryByTestId,
        findByText,
      } = await setup({ mockResponse, widgetOptions: { features: { autoFocus: true } } });

      await findByTestId('identifier') as HTMLInputElement;
      await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByText('Sign in', { selector: 'button' });

      await user.click(submitButton);
      await findByText(/We found some errors./);
      const identifierError = await findByTestId('identifier-error');
      const passwordError = await findByTestId('credentials.passcode-error');

      expect(identifierError.textContent).toEqual('This field cannot be left blank');
      expect(container).toMatchSnapshot();
      expect(passwordError.textContent).toEqual('This field cannot be left blank');
      expect(authClient.options.httpRequestClient).not.toHaveBeenCalled();

      const identifierField = await findByTestId('identifier') as HTMLInputElement;
      await user.type(identifierField, 'someuser@okta1.com');

      expect(queryByTestId('identifier-error')).toBeNull();
      expect((await findByTestId('credentials.passcode-error')).textContent).toBe('This field cannot be left blank');
    });

    it('should type in field, then clear field to view field level error', async () => {
      const {
        user,
        findByTestId,
        queryByTestId,
        findByText,
      } = await setup({ mockResponse, widgetOptions: { features: { autoFocus: true } } });

      const identifierEle = await findByTestId('identifier') as HTMLInputElement;
      await findByTestId('credentials.passcode') as HTMLInputElement;
      await findByText('Sign in', { selector: 'button' });
      expect(queryByTestId('identifier-error')).toBeNull();
      expect(queryByTestId('credentials.passcode-error')).toBeNull();

      await user.type(identifierEle, 'aaa');
      await user.clear(identifierEle);
      await user.tab();

      const identifierError = await findByTestId('identifier-error');
      expect(identifierError.textContent).toEqual('This field cannot be left blank');
      expect(queryByTestId('credentials.passcode-error')).toBeNull();
    });

    it('fails client side validation with no inputs', async () => {
      const {
        authClient,
        user,
        container,
        findByTestId,
        findByText,
        queryByTestId,
      } = await setup({ mockResponse, widgetOptions: { features: { autoFocus: true } } });
      let identifierError;
      let passwordError;

      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByText('Sign in', { selector: 'button' });

      // empty username & empty password
      await user.click(submitButton);
      await findByText(/We found some errors./);
      identifierError = await findByTestId('identifier-error');
      expect(identifierError.textContent).toEqual('This field cannot be left blank');
      expect(container).toMatchSnapshot();

      passwordError = await findByTestId('credentials.passcode-error');
      expect(passwordError.textContent).toEqual('This field cannot be left blank');
      expect(authClient.options.httpRequestClient).not.toHaveBeenCalled();

      // add username - updates field level error
      await user.type(usernameEl, 'testuser@okta.com');
      await user.click(submitButton);
      await waitFor(async () => {
        passwordError = await findByTestId('credentials.passcode-error');
        expect(passwordError.textContent).toEqual('This field cannot be left blank');
        identifierError = queryByTestId('identifier-error');
        expect(identifierError).toBeNull();
        expect(authClient.options.httpRequestClient).not.toHaveBeenCalled();
      });

      // add password - clears error and send payload
      await user.type(passwordEl, 'fake-password');
      await user.click(submitButton);
      await waitFor(async () => {
        expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
          ...createAuthJsPayloadArgs('POST', 'idp/idx/identify', {
            identifier: 'testuser@okta.com',
            credentials: {
              passcode: 'fake-password',
            },
          }),
        );
        identifierError = queryByTestId('identifier-error');
        expect(identifierError).toBeNull();
        passwordError = queryByTestId('credentials.passcode-error');
        expect(passwordError).toBeNull();
      });
    });
  });

  describe('sends correct payload', () => {
    it('with all required fields', async () => {
      const {
        authClient, user, findByTestId, findByText,
      } = await setup({ mockResponse });

      const titleElement = await findByText('Sign In', { selector: 'h2' });
      await waitFor(() => expect(titleElement).toHaveFocus());
      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByText('Sign in', { selector: 'button' });

      await user.type(usernameEl, 'testuser@okta.com');
      expect(usernameEl.value).toEqual('testuser@okta.com');
      await user.type(passwordEl, 'fake-password');
      expect(passwordEl.value).toEqual('fake-password');
      await user.click(submitButton);

      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/identify', {
          identifier: 'testuser@okta.com',
          credentials: {
            passcode: 'fake-password',
          },
        }),
      );
    });

    it('should modify username for PRIMARY_AUTH operation when transformUsername function is defined', async () => {
      const {
        authClient, user, findByTestId, findByText,
      } = await setup({
        mockResponse,
        widgetOptions: {
          transformUsername: (username: string, operation) => (
            operation === 'PRIMARY_AUTH' ? username.split('@')[0] : username
          ),
        },
      });

      const titleElement = await findByText('Sign In', { selector: 'h2' });
      await waitFor(() => expect(titleElement).toHaveFocus());
      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByText('Sign in', { selector: 'button' });
      const mockUsername = 'testuser@okta1.com';

      await user.type(usernameEl, mockUsername);
      expect(usernameEl.value).toEqual(mockUsername);
      await user.type(passwordEl, 'fake-password');
      expect(passwordEl.value).toEqual('fake-password');
      await user.click(submitButton);

      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/identify', {
          identifier: 'testuser',
          credentials: { passcode: 'fake-password' },
        }),
      );
    });

    it('should not modify username for when operation is PRIMARY_AUTH as defined in transformUsername function', async () => {
      const {
        authClient, user, findByTestId, findByText,
      } = await setup({
        mockResponse,
        widgetOptions: {
          transformUsername: (username: string, operation) => (
            operation === 'PRIMARY_AUTH' ? username : 'DUMMY_USERNAME'
          ),
        },
      });

      const titleElement = await findByText('Sign In', { selector: 'h2' });
      await waitFor(() => expect(titleElement).toHaveFocus());
      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByText('Sign in', { selector: 'button' });
      const mockUsername = 'testuser@okta1.com';

      await user.type(usernameEl, mockUsername);
      expect(usernameEl.value).toEqual(mockUsername);
      await user.type(passwordEl, 'fake-password');
      expect(passwordEl.value).toEqual('fake-password');
      await user.click(submitButton);

      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/identify', {
          identifier: mockUsername,
          credentials: { passcode: 'fake-password' },
        }),
      );
    });

    it('should not remove padded spaces from password', async () => {
      const {
        authClient, user, findByTestId, findByText,
      } = await setup({ mockResponse });

      const titleElement = await findByText('Sign In', { selector: 'h2' });
      await waitFor(() => expect(titleElement).toHaveFocus());
      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByText('Sign in', { selector: 'button' });

      const password = '   fake-password   ';
      await user.type(usernameEl, 'testuser@okta.com');
      expect(usernameEl.value).toEqual('testuser@okta.com');
      await user.type(passwordEl, password);
      expect(passwordEl.value).toEqual(password);
      await user.click(submitButton);

      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/identify', {
          identifier: 'testuser@okta.com',
          credentials: {
            passcode: password,
          },
        }),
      );
    });

    it('with required fields + optional fields', async () => {
      const {
        authClient, user, findByTestId, findByText,
      } = await setup({ mockResponse });

      const titleElement = await findByText('Sign In', { selector: 'h2' });
      await waitFor(() => expect(titleElement).toHaveFocus());
      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const rememberMeEl = await findByTestId('rememberMe');
      const submitButton = await findByText('Sign in', { selector: 'button' });

      await user.type(usernameEl, 'testuser@okta.com');
      expect(usernameEl.value).toEqual('testuser@okta.com');
      await user.type(passwordEl, 'fake-password');
      expect(passwordEl.value).toEqual('fake-password');
      await user.click(rememberMeEl);
      await user.click(submitButton);

      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/identify', {
          identifier: 'testuser@okta.com',
          credentials: {
            passcode: 'fake-password',
          },
          rememberMe: true,
        }),
      );
    });
  });

  it('should only send one api request when submit button is double clicked', async () => {
    const {
      authClient, user, findByTestId, findByText,
    } = await setup({ mockResponse });

    const titleElement = await findByText('Sign In', { selector: 'h2' });
    await waitFor(() => expect(titleElement).toHaveFocus());
    const usernameEl = await findByTestId('identifier') as HTMLInputElement;
    const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
    const submitButton = await findByText('Sign in', { selector: 'button' });

    await user.type(usernameEl, 'testuser@okta.com');
    expect(usernameEl.value).toEqual('testuser@okta.com');
    await user.type(passwordEl, 'fake-password');
    expect(passwordEl.value).toEqual('fake-password');
    await user.dblClick(submitButton);

    expect(authClient.options.httpRequestClient).toHaveBeenCalledTimes(1);
  });

  it('should only send one api request when forgot password link is double clicked', async () => {
    const {
      findByText,
      user,
      authClient,
    } = await setup({ mockResponse });

    const cancelLink = await findByText(/Forgot password?/);

    await user.dblClick(cancelLink);
    expect(authClient.options.httpRequestClient).toHaveBeenCalledTimes(1);
  });

  it('identifier and password fields should be ltr even when a rtl language is set', async () => {
    const {
      findByTestId,
    } = await setup({ mockResponse, widgetOptions: { language: 'ar' } });

    const usernameEl = await findByTestId('identifier') as HTMLInputElement;
    const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;

    expect(usernameEl.parentElement?.getAttribute('dir')).toBe('ltr');
    expect(passwordEl.parentElement?.getAttribute('dir')).toBe('ltr');
  });
});
