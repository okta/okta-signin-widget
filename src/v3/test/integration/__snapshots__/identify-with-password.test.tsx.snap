// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`identify-with-password renders form 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <span>
      <main
        class="auth-container main-container mainViewContainer MuiBox-root emotion-1"
        id="okta-sign-in"
      >
        <div
          class="siwContainer MuiBox-root emotion-2"
        >
          <div
            class="okta-sign-in-header auth-header siwHeader"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 emotion-3"
            />
          </div>
          <div
            class="MuiBox-root emotion-4"
          >
            <form
              class="o-form"
              data-se="form"
              novalidate=""
            >
              <div
                class="MuiBox-root emotion-5"
              >
                <div
                  class="MuiBox-root emotion-6"
                >
                  <div
                    class="MuiBox-root emotion-7"
                  >
                    <h2
                      class="MuiTypography-root MuiTypography-h3 emotion-8"
                    >
                      Sign In
                    </h2>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <div
                    class="MuiBox-root emotion-10"
                  >
                    <label
                      class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary emotion-11"
                      for="identifier"
                    >
                      Username
                    </label>
                    <div
                      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth emotion-12"
                    >
                      <input
                        aria-invalid="false"
                        autocomplete="username"
                        class="MuiOutlinedInput-input MuiInputBase-input emotion-13"
                        data-se="identifier"
                        id="identifier"
                        name="identifier"
                        type="text"
                      />
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline emotion-14"
                      >
                        <legend
                          class="emotion-15"
                        >
                          <span
                            class="notranslate"
                          >
                            ​
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <div
                    class="MuiBox-root emotion-10"
                  >
                    <div
                      class="MuiBox-root emotion-10"
                    >
                      <label
                        class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary emotion-11"
                        for="credentials.passcode"
                      >
                        Password
                      </label>
                      <div
                        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedEnd emotion-20"
                      >
                        <input
                          aria-invalid="false"
                          autocomplete="current-password"
                          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd emotion-21"
                          data-se="credentials.passcode"
                          id="credentials.passcode"
                          name="credentials.passcode"
                          type="password"
                        />
                        <div
                          class="MuiInputAdornment-root MuiInputAdornment-positionEnd emotion-22"
                        >
                          <button
                            aria-label="toggle password visibility"
                            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium emotion-23"
                            data-mui-internal-clone-element="true"
                            tabindex="0"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-24"
                              data-testid="VisibilityIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                              />
                            </svg>
                          </button>
                        </div>
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline emotion-14"
                        >
                          <legend
                            class="emotion-15"
                          >
                            <span
                              class="notranslate"
                            >
                              ​
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <label
                    class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd emotion-28"
                  >
                    <span
                      class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root emotion-29"
                    >
                      <input
                        class="PrivateSwitchBase-input emotion-30"
                        data-se="rememberMe"
                        data-se-for-name="rememberMe"
                        id="rememberMe"
                        name="rememberMe"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-24"
                        data-testid="CheckBoxOutlineBlankIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label emotion-32"
                    >
                      Keep me signed in
                    </span>
                  </label>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <button
                    class="MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButton-disableElevation MuiButton-fullWidth MuiButtonBase-root  emotion-34"
                    data-type="save"
                    tabindex="0"
                    type="submit"
                  >
                    Sign in
                  </button>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <button
                    class="MuiButton-root MuiButton-floating MuiButton-floatingPrimary MuiButton-sizeMedium MuiButton-floatingSizeMedium MuiButton-disableElevation MuiButtonBase-root  emotion-36"
                    data-se="forgot-password"
                    tabindex="0"
                    type="button"
                  >
                    Forgot password?
                  </button>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <button
                    class="MuiButton-root MuiButton-floating MuiButton-floatingPrimary MuiButton-sizeMedium MuiButton-floatingSizeMedium MuiButton-disableElevation MuiButtonBase-root  emotion-36"
                    data-se="enroll"
                    tabindex="0"
                    type="button"
                  >
                    Register
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>
      </main>
    </span>
  </div>
</div>
`;

exports[`identify-with-password renders the loading state first 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <span>
      <main
        class="auth-container main-container mainViewContainer MuiBox-root emotion-1"
        id="okta-sign-in"
      >
        <div
          class="siwContainer MuiBox-root emotion-2"
        >
          <div
            class="okta-sign-in-header auth-header siwHeader"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 emotion-3"
            />
          </div>
          <div
            class="MuiBox-root emotion-4"
          >
            <div
              class="MuiBox-root emotion-5"
            >
              <span
                aria-label="Loading..."
                aria-valuetext="Loading..."
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary emotion-6"
                role="progressbar"
                style="width: 1.14285714rem; height: 1.14285714rem;"
              >
                <svg
                  class="MuiCircularProgress-svg emotion-7"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate emotion-8"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="18"
                    stroke-width="8"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
      </main>
    </span>
  </div>
</div>
`;

exports[`identify-with-password sends correct payload fails client side validation with no inputs 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <span>
      <main
        class="auth-container main-container mainViewContainer MuiBox-root emotion-1"
        id="okta-sign-in"
      >
        <div
          class="siwContainer MuiBox-root emotion-2"
        >
          <div
            class="okta-sign-in-header auth-header siwHeader"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 emotion-3"
            />
          </div>
          <div
            class="MuiBox-root emotion-4"
          >
            <div
              class="MuiBox-root emotion-5"
            >
              <div
                class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-infoboxError MuiAlert-infobox emotion-6"
                role="alert"
              >
                <div
                  class="MuiAlert-icon emotion-7"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-8"
                    data-testid="ErrorOutlineIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiAlert-message emotion-9"
                >
                  We found some errors. Please review the form and make corrections.
                </div>
              </div>
            </div>
            <form
              class="o-form"
              data-se="form"
              novalidate=""
            >
              <div
                class="MuiBox-root emotion-10"
              >
                <div
                  class="MuiBox-root emotion-11"
                >
                  <div
                    class="MuiBox-root emotion-12"
                  >
                    <h2
                      class="MuiTypography-root MuiTypography-h3 emotion-13"
                    >
                      Sign In
                    </h2>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-11"
                >
                  <div
                    class="MuiBox-root emotion-15"
                  >
                    <label
                      class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary emotion-16"
                      for="identifier"
                    >
                      Username
                    </label>
                    <div
                      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth emotion-17"
                    >
                      <input
                        aria-invalid="true"
                        autocomplete="username"
                        class="MuiOutlinedInput-input MuiInputBase-input emotion-18"
                        data-se="identifier"
                        id="identifier"
                        name="identifier"
                        type="text"
                      />
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline emotion-19"
                      >
                        <legend
                          class="emotion-20"
                        >
                          <span
                            class="notranslate"
                          >
                            ​
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                    <p
                      class="MuiFormHelperText-root Mui-error emotion-21"
                      data-se="identifier-error"
                    >
                      This field cannot be left blank
                    </p>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-11"
                >
                  <div
                    class="MuiBox-root emotion-15"
                  >
                    <div
                      class="MuiBox-root emotion-15"
                    >
                      <label
                        class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary emotion-16"
                        for="credentials.passcode"
                      >
                        Password
                      </label>
                      <div
                        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-adornedEnd emotion-26"
                      >
                        <input
                          aria-invalid="true"
                          autocomplete="current-password"
                          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd emotion-27"
                          data-se="credentials.passcode"
                          id="credentials.passcode"
                          name="credentials.passcode"
                          type="password"
                        />
                        <div
                          class="MuiInputAdornment-root MuiInputAdornment-positionEnd emotion-28"
                        >
                          <button
                            aria-label="toggle password visibility"
                            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium emotion-29"
                            data-mui-internal-clone-element="true"
                            tabindex="0"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-30"
                              data-testid="VisibilityIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                              />
                            </svg>
                          </button>
                        </div>
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline emotion-19"
                        >
                          <legend
                            class="emotion-20"
                          >
                            <span
                              class="notranslate"
                            >
                              ​
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                    <p
                      class="MuiFormHelperText-root Mui-error emotion-21"
                      data-se="credentials.passcode-error"
                    >
                      This field cannot be left blank
                    </p>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-11"
                >
                  <label
                    class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd emotion-35"
                  >
                    <span
                      class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root emotion-36"
                    >
                      <input
                        class="PrivateSwitchBase-input emotion-37"
                        data-se="rememberMe"
                        data-se-for-name="rememberMe"
                        id="rememberMe"
                        name="rememberMe"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-30"
                        data-testid="CheckBoxOutlineBlankIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label emotion-39"
                    >
                      Keep me signed in
                    </span>
                  </label>
                </div>
                <div
                  class="MuiBox-root emotion-11"
                >
                  <button
                    class="MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButton-disableElevation MuiButton-fullWidth MuiButtonBase-root  emotion-41"
                    data-type="save"
                    tabindex="0"
                    type="submit"
                  >
                    Sign in
                  </button>
                </div>
                <div
                  class="MuiBox-root emotion-11"
                >
                  <button
                    class="MuiButton-root MuiButton-floating MuiButton-floatingPrimary MuiButton-sizeMedium MuiButton-floatingSizeMedium MuiButton-disableElevation MuiButtonBase-root  emotion-43"
                    data-se="forgot-password"
                    tabindex="0"
                    type="button"
                  >
                    Forgot password?
                  </button>
                </div>
                <div
                  class="MuiBox-root emotion-11"
                >
                  <button
                    class="MuiButton-root MuiButton-floating MuiButton-floatingPrimary MuiButton-sizeMedium MuiButton-floatingSizeMedium MuiButton-disableElevation MuiButtonBase-root  emotion-43"
                    data-se="enroll"
                    tabindex="0"
                    type="button"
                  >
                    Register
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>
      </main>
    </span>
  </div>
</div>
`;
